docs/Backend_tutorials/features/graphql.md

```diff
@@ -1,5 +1,139 @@
+
+We are using `Mutations::Base` for all our mutations. This class is defined in `app/api/easy_graphql/mutations/base.rb`.
+
+Place mutations into `app/api/easy_graphql/mutations` directory. Name convention of mutation should describe "what mutation does" OR if its change entity, just use entity name **with action suffix**. For example `IssueCreate` or `IssueUpdate`.
+
+> [!CAUTION]
+> NEVER ever do more actions in one mutations.
+> 
+> Its forbidden to combine `create` and `update` in one mutation ! Always use separate mutations for each logical action. 
+
+### Error handling
+https://graphql-ruby.org/mutations/mutation_classes.html#error-handling
+
+We are using `Types::Errors` for all our errors. This class is defined in `app/api/easy_graphql/types/error.rb`. This class is used for defining errors of the mutation. Define `errors` on your mutation as new field:
+
+```ruby
+field :errors, [Types::Error], null: true
+```
+
+> [!WARNING]
+> `_errors` must be always array of `Types::Error` !
+
+In mutation there is method `error` which return `Errors` object for store validation errors. (see easy_engines/easy_extensions/app/api/easy_graphql/mutations/errors.rb)
+
+Object have public methods: 
+
+```ruby title="easy_engines/easy_extensions/app/api/easy_graphql/mutations/errors.rb" lineNumbers
+# Add an error to the errors array.
+# @param attribute [Symbol, String] attribute associated with the error
+# @param messages [Array<String>] error messages
+def add_error(attribute:, messages:)
+  errors << { attribute:, messages: }
+end
+
+# Add errors from an entity to the errors array.
+# @param entity [ActiveRecord::Base] entity to add errors from
+def add_entity_errors(entity)
+  entity_errors = EasyExtensions::ParseActiveRecordErrors.call(entity)
+  errors.concat(entity_errors)
+end
+```
+
+So when you are updating standard ActiveRecord object in mutation, use 
+```ruby title="Example of resolve method with validation"
+entity.init_journal(User.current, "My changes")
+errors.add_entity_errors(entity) unless entity.save
+
+{ entity:, errors: errors }
+```
+
+### Best-practice
+
+1. **Use `resolve` method for changing data.**
+2. **Use `field` errors aside your returns.**
+3. **Always return `errors` field.** If there is no error, return empty array.
+4. **Use `argument` method for defining fields of the type.** Do not forget specify `required` option for argument.
+5. If you find specify entity, use `find` method for fetching entity. It raise `ActiveRecord::RecordNotFound` if entity not found = This is correct.
+6. Do not forget `init_journal` if entity is journalized.
+7. Use `safe_attributes` for assign attributes to entity.
+8. If you create new entity and its not valid, return `errors` and `nil` for entity.
```

Comment:
What enums. Now we have two definition and usage.
1) modul definition sigle VS plural (status/statuses) ? I prefer plural.
2) Some graphql type return as value Enum type but other return only string and Enum type use only for agruments.
easy_engines/easy_ai/app/api/easy_graphql/types/easy_ai/feedback.rb VS   plugins/easyproject/easy_plugins/easy_scrum_boards/api/easy_graphql/types/easy_agile_sprint.rb  (status is enum)

Reply:
Thanks for feedback. Based on internal discussion:

1. We want follow ruby-graphql best-practice based on their docs => So use `singular` form of enums field (because mentioned enum represent specify field - "status" for example)
2. definitely agreee - use enums in fields and arguments too.